pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1885,6 +1885,17 @@ def test_dtype_on_merged_different(self, change, join_type, left, right):
         )
         tm.assert_series_equal(result, expected)
 
+        # GH 43734 Avoid the use of `assign` with multiindex
+        right.columns = MultiIndex.from_tuples([("lvl0", x) for x in right.columns])
+        left.columns = MultiIndex.from_tuples([("lvl0", x) for x in left.columns])
```

Comment:
this doesn't appear to actually be testing the OP, e.g. that's where the dtype is str vs StringDtype(). can you parameterize and show the dataframe construction.